### PR TITLE
provider/openstack: Load balancer onDemand Caching

### PR DESCRIPTION
**Pull request by derekolk**
closes https://github.com/spinnaker/spinnaker/issues/1029

### CODE REVIEW

**Comment by derekolk** [comment]
@spinnaker/reviewers @lwander

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+   * Builds on demand cache result based upon the cache result.
+   * @param onDemandAgentType
+   * @param cacheResult
+   * @param namespace
+   * @param key
+   * @param providerCache
+   * @return
+   */
+  OnDemandResult buildOnDemandCache(Object cachedItem
+                                    , String onDemandAgentType
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: I haven't seen this argument alignment style elsewhere, could we keep in consistent with 
``` 
...(type arg, 
   type arg,
   type arg)
```

Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: Agreed ... will change.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    if (!cachedItem) {
+      // Evict this cached item if it no longer exists.
+      result.evictions[namespace] << key
+    }
+
+    result
+  }
+
+  /**
+   * Resolves a key that contains a wildcard but still returns a unique result.
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: when passed a key that does not resolve a unique result this method does not follow this contract. I'd recommend either throwing an exception, or explaining that it only resolves keys to unique results when possible.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 3]
```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2016 Target, Inc.
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: Nice work centralizing a lot of this shared on demand code.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy [hunk 4]
```diff
@@ -50,10 +50,22 @@ class OpenstackNetworkingV2Provider implements OpenstackNetworkingProvider, Open
...
+
+  List<? extends LbPool> getInternalAllLoadBalancerPools(final String region, Map<String, String> filterParameters = null) {
+    handleRequest {
+      getRegionClient(region).networking().loadbalancers().lbPool().list(filterParameters)
+    }
+  }
+
+  @Override
+  LbPool getLoadBalancerPoolByName(final String region, final String loadBalancerName) {
+    List<LbPool> pools = getInternalAllLoadBalancerPools(region, ['name':loadBalancerName])
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy]: Quotes aren't needed in groovy for a string key in a hashmap.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy (hunk 2 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: I will update it to throw an exception ... Would like to know if it isn't uniquely resolving.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy [hunk 5]
```diff
@@ -45,80 +53,163 @@ import static com.netflix.spinnaker.clouddriver.openstack.cache.Keys.Namespace.V
...
+      if (shouldUseOnDemandData(cacheResultBuilder, loadBalancerKey)) {
+        moveOnDemandDataToNamespace(objectMapper, typeReference, cacheResultBuilder, loadBalancerKey)
+      } else {
+        //health monitors get looked up
+        Set<HealthMonitor> healthMonitors = pool.healthMonitors?.collect { healthId ->
+          clientProvider.getHealthMonitor(region, healthId)
+        }?.toSet()
+
+        //vips cached
+        Map<String, Object> vipMap = providerCache.get(VIPS.ns, Keys.getVipKey(pool.vipId, accountName, region))?.attributes
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: Looking up the cache data and storing the results a second time is likely the wrong approach here. The better (more efficient way) is to simply store the `vip`s, `ip`s, `port`s, etc... as in the `.relationships` field of cache data object, and reconstruct the full load balancer object in the `View` class. This way you don't store duplicate data, and don't have extra redis traffic during the cache cycle.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy (hunk 4 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy]: will change.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy (hunk 3 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: Thanks 👍

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy (hunk 5 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: Thanks for the heads up ... We are in the process of rewriting the majority of load balancer functionality to support lbaasv2 and remove lbaasv1.  I will add this comment to our caching ticket.

**Comment by derekolk** [comment]
@lwander Updated based upon your comments.

**Comment by anotherchrisberry** [comment]
needs a rebase

**Comment by derekolk** [comment]
@anotherchrisberry rebased

**Pull request merged**